Sources/NIO/BaseSocketChannel.swift

```diff
@@ -611,19 +786,23 @@ class BaseSocketChannel<T: BaseSocket>: SelectableChannel, ChannelCore {
                     return
                 }
             } else {
-                pipeline.fireErrorCaught0(error: err)
+                self.pipeline.fireErrorCaught0(error: err)
             }
 
             // Call before triggering the close of the Channel.
-            pipeline.fireChannelReadComplete0()
+            if self.lifecycleManager.isActive {
```

Comment:
Is this right? It's not clear to me that the contract of `readComplete` requires that it only fire on active channels. If it does, is that contract in tension with the contract that `readComplete` always fires after `channelRead`?

Reply:
I think in general it should only fire when the channel is active.

Reply:
@Lukasa, chatted to @normanmaurer and added a test which always gets into this situation by doing

1. one successful read
2. one read that fails with `ECONNRESET`
3. in the `errorCaught` calls `ctx.close` to close the channel
4. sees `channelInactive` (from the close)
5. does _not_ ever see `channelReadComplete`.